janusgraph-core/src/main/java/org/janusgraph/graphdb/types/vertices/JanusGraphSchemaVertex.java

```diff
@@ -176,11 +175,9 @@ public SchemaStatus getStatus() {
     @Override
     public IndexType asIndexType() {
         Preconditions.checkArgument(getDefinition().containsKey(TypeDefinitionCategory.INTERNAL_INDEX),"Schema vertex is not a type vertex: [%s,%s]", longId(), name());
-        if (getDefinition().<Boolean>getValue(TypeDefinitionCategory.INTERNAL_INDEX)) {
-            return new CompositeIndexTypeWrapper(this);
-        } else {
-            return new MixedIndexTypeWrapper(this);
-        }
+        return getDefinition().<Boolean>getValue(TypeDefinitionCategory.INTERNAL_INDEX) ?
+            new CompositeIndexTypeWrapper(this) :
+            new MixedIndexTypeWrapper(this);
```

Comment:
As explained here https://github.com/JanusGraph/janusgraph/pull/2688#discussion_r663368800 it looks like we can actually cache the returning `IndexType`. The example implementation is here: https://github.com/porunov/janusgraph/commit/ce9191ef302093c9ae9b13ca5d13dab1cbe8e4ad (all tests passed).
That said, as I'm not sure 100% here that it's OK to cache the resulting object. Thus, I didn't do it in this PR. That said, I think we should check it and possibly change the implementation of this method to actually cache the result instead of returning a new instance every time (probably in the next PR if we can prove that it's OK to cache the result).